src/_pytest/cacheprovider.py

```diff
@@ -258,7 +258,7 @@ def pytest_collection_modifyitems(self, session, config, items):
             self._report_status = "no previously failed tests, "
             if self.config.getoption("last_failed_no_failures") == "none":
                 self._report_status += "deselecting all items."
-                config.hook.pytest_deselected(items=items)
+                config.hook.pytest_deselected(items=items[:])
```

Comment:
Good catch!

I wonder if it is a cache provider bug though... I mean, should every caller of `pytest_deselected(items)` always send a copy over? 🤔 

Sure, `pytest_deselected` is kind of a weird hook, but I thought I would raise this.

Reply:
Not sure myself, but here it is clear that the list is cleared just below at least.

FWIW, I've found an issue with auto-deselecting of tests when I was passing a `set` there (and trying to `items[:]` it).
As for the terminal plugin only the length is relevant anyway I think (#6409), but thought to keep `stats` there like before.

Reply:
for sanity we should actually fail  if session.items is passed by accident

this is a great and horrifying catch

the typical/presumed usage is collecting keep and drop in new lists, then passing drop after changing items, so the error is a honest misstake

Reply:
@RonnyPfannschmidt 
> for sanity we should actually fail if session.items is passed by accident

Makes sense, please consider creating a new issue for it.

> the typical/presumed usage is collecting keep and drop in new lists, then passing drop after changing items, so the error is a honest misstake

I could not really parse this.